tokio-trace-core/src/dispatcher.rs

```diff
@@ -35,11 +36,16 @@ impl Dispatch {
     where
         F: FnOnce(&Dispatch) -> T,
     {
-        if let Some(c) = Span::current().dispatch() {
-            f(c)
-        } else {
-            CURRENT_DISPATCH.with(|current| f(&*current.borrow()))
+        // If we try to access the current dispatcher while it's being
+        // dropped, `LocalKey::with` would panic, causing a double panic.
+        // However, we can't use `try_with` as we still need to invoke `f`,
+        // which would be captured by the closure.
```

Comment:
This is a largely unrelated change, but it's something I stumbled across while testing --- sometimes test failures would drop something that called this function, and then we'd panic.

Reply:
Could this be done using `LocalKey::try_with`?

Reply:
@carllerche That was the first thing I tried. However, since `f` is a `FnOnce`, something like 
```rust
CURRENT_DISPATCH.try_with(|current| {
    f(&*current.borrow())
}).unwrap_or_else(||  f(&Dispatch::none())
```
because `f` has to be moved into the closure passed to `try_with` (as it's a `FnOnce` and must be moved to be invoked).

Reply:
The issue is there are more cases than just panicking that could trigger a `try` panic...